src/libraries/System.Private.CoreLib/src/System/Globalization/TextInfo.cs

```diff
@@ -200,14 +203,81 @@ internal void ChangeCaseToUpper(ReadOnlySpan<char> source, Span<char> destinatio
             ChangeCaseCommon<ToUpperConversion>(ref MemoryMarshal.GetReference(source), ref MemoryMarshal.GetReference(destination), source.Length);
         }
 
+        private unsafe void ChangeCaseCommon_Vector128<TConversion>(ref char source, ref char destination, int charCount)
+            where TConversion : struct
+        {
+            Debug.Assert(charCount >= Vector128<ushort>.Count);
+            Debug.Assert(Vector128.IsHardwareAccelerated);
+
+            // JIT will treat this as a constant in release builds
+            bool toUpper = typeof(TConversion) == typeof(ToUpperConversion);
+
+            ref ushort src = ref Unsafe.As<char, ushort>(ref source);
+            ref ushort dst = ref Unsafe.As<char, ushort>(ref destination);
+
+            nuint lengthU = (nuint)charCount;
+            nuint lengthToExamine = lengthU - (nuint)Vector128<ushort>.Count;
+            nuint i = 0;
+            do
+            {
+                Vector128<ushort> vec = Vector128.LoadUnsafe(ref src, i);
+                if (!Utf16Utility.AllCharsInVector128AreAscii(vec))
+                {
+                    goto NON_ASCII;
+                }
+                vec = toUpper ?
+                    Utf16Utility.Vector128AsciiToUppercase(vec) :
+                    Utf16Utility.Vector128AsciiToLowercase(vec);
+                vec.StoreUnsafe(ref dst, i);
+
+                i += (nuint)Vector128<ushort>.Count;
+            } while (i <= lengthToExamine);
```

Comment:
```suggestion
            } while (i < lengthToExamine);
```
Can safe one iteration of the loop.
If `i` is exactly one vector away from end, then with `<=` another iteration is done, as well the remainder processed again (vectorized). This can be avoided by using `<`.
All other cases behave the same whether `<=` or `<`.

Reply:
@gfoidl can you elaborate what exactly we save here? we either do one more iteration inside the loop or enter "handle trailing elements" block - why the latter is better?

Reply:
You remove the if below and just process the remainder anyway. Length check is already done before.
I'm on mobile, so sorry for being not more descriptive.

Reply:
what about when the source length is exactly 8 chars? so we have only 1 vector to process

Reply:
Ey, in this method is no initial length check before the loop (I over-read that). So maybe it's best to keep the code as is?
For cases like this I don't know what is the best way anyway. For exactly 8 chars it would be done twice. But for multiples of 8 the branch can be removed. It's hard to measure with todays branch-predictors, but my theory for this is https://github.com/dotnet/runtime/pull/76337#discussion_r983448480. Here we would penalize all cases bit, just to reward 8. Again: I don't know if it's worth it, and with micro-benchmarks it's hard to measure.

Reply:
@gfoidl exactly, the penalty for 8 chars will be quite noticeable while for large inputs it will be close to noise so I'd probably leave it as is. We generally try to prioritize small inputs over large ones